Revamp docs, provide better composition utilities (#936)
NullVoxPopuli/ember-resources: docs/docs/resources.md

# Resources

_this document has been adapted/copied[^copying] from the Starbeam[^starbeam] documentation_

> This is a highlevel introduction to Resources, and how to use them.

Reviewer: ```suggestion
This is a high-level introduction to Resources, and how to use them.
```

Author: thanks!

For how to integrate Resources in to Ember (Components, etc), see ['./ember.md]('./ember.md'); 

> In addition to the live demos accompanying each code snippet, all code snippets will have their starbeam counterparts below them, so that folks can see how similar the libraries are.

Reviewer: ```suggestion
In addition to the live demos accompanying each code snippet, all code snippets will have their Starbeam counterparts below them, so that folks can see how similar the libraries are.
```

Author: thanks!

<details>
<summary>Resources Convert Processes Into Values</summary>

Typically, a resource converts an imperative, stateful process, such as an asynchronous request or a ticking timer, into a reactive value.

That allows you to work with a process just like you'd work with any other reactive value.

This is a very powerful capability, because it means that adding cleanup logic to an existing reactive value doesn't change the code that works with the value.

The only thing that changes when you convert a reactive value into a resource is that it must be instantiated with an owner. The owner defines the resource's lifetime. Once you've instantiated a resource, the value behaves like any other reactive value.

</details>

## A Very Simple Resource

To illustrate the concept, let's create a simple resource that represents the current time.

```js 
import { cell, resource } from "ember-resources";
 
export const Now = resource(({ on }) => {
  const now = cell(Date.now());
 
  const timer = setInterval(() => {
    now.set(Date.now());
  });
 
  on.cleanup(() => {
    clearInterval(timer);
  });
 
  return now;
});
```

> To see this code in action, [checkout the live demo](https://limber.glimdown.com/edit?c=MQAggiDKAuD2AOB3AhtAxgCwFBbbAJgKYgBmhAdmsdMgOYDOIay5IARsQK72H4hwh8ASwBOhNNAA2ATxBjyREU1gBbeLHIVo9ADRZkjRIUmSQBuRUX8MxPERD1yQ%2BPELRrxSUIButgn5VCRhIRVQA6HAB5eGghDXoALhwAWhAAAy9fNJBUgH15RUYhVnhJZCpcrFSMn2J4MW8hQkRs-MtCESKSsqodEFLy4toPB1hOESozEmgOkcziAo7K6vn%2BhqbEdmNYFpA2hQ6u-p7CPoG0IbnaiwOlDkkdkfoxicJKrDTP2gArRnmsIRqWAidwAbwsz3GvSYxlMAF9SKEVCAAOSEFQcETJMSQ170FEAbhwhAAHuoQcpyPR3AA5R4AXghLyoAAoWeCNCA4QBKED0gB8IFBWBAlOpIHIDJhJhZABFUIQwpLECzudyiSARWL3LFAkpGTxoABJcgzETeZCSNm8gVCrWi5VhQ1yhVKnaq9VankarUaMJoSSEFiceDWvmC4Wi0UBoMiE1mi1W3UdT2i71arViaDjVjKonpnAAHhmajKM35WqN7iEiRAxcBhH5oNBdMQcLhhYA9MmK12SwNyx9PkA&format=glimdown).

Reviewer: This demo doesn't work:

![image](https://github.com/NullVoxPopuli/ember-resources/assets/9835/5a33ab06-1b2f-4009-be2a-332829d62786)

Also, it's unclear to me what the "options" it is talking about are.

Author: demo will work when the PR is merged :sweat_smile:  -- because it implements the needed feature where Cells are directly renderable (via the smallest helper manager ever :tada: )

Author: > Also, it's unclear to me what the "options" it is talking about are.

hm, yeah, it's a little orthogonal -- but it's how to use code fences on https://limber.glimdown.com -- maybe that should be removed and extracted to a separate set of docs

    clearInterval(timer);
  });
 
  return now;
});
```

</details>

> **:bulb:** <br> 
> A resource's return value is a reactive value. If your resource represents a single cell, it's fine to return it directly. It's also common to return a function which returns reactive data -- that depends on reactive state that you created inside the resource constructor.

When you use the `Now` resource in a component, it will automatically get its lifetime linked to that component. In this case, that means that the interval will be cleaned up when the component is destroyed.

The `resource` function creates a resource Constructor. A resource constructor:

1. Sets up internal reactive state that changes over time.
2. Sets up the external process that needs to be cleaned up.
3. Registers the cleanup code that will run when the resource is cleaned up.
4. Returns a reactive value that represents the current state of the resource as a value.

In this case:

| internal state | external process | cleanup code | return value |
| ---- | ---- | ---- | ---- |
| `Cell<number>` | `setInterval` | `clearInterval` | `Cell<number>` |

<details><summary>Resource's values are immutable</summary>

> When you return a reactive value from a resource, it will always behave like a generic, immutable reactive value. This means that if you return a `Cell` from a resource, the resource's value will have `.current` and `.read()`, but not `.set()`, `.update()` or other cell-specific methods.

Reviewer: ```suggestion
When you return a reactive value from a resource, it will always behave like a generic, immutable reactive value. This means that if you return a `cell` from a resource, the resource's value will have `.current` and `.read()`, but not `.set()`, `.update()` or other cell-specific methods.
```

Looks like this was copied from Starbeam, given the capitalization. Are `current` and `read` still the method names?

Author: yeah, `current`, `read`, `set`, and `update` all match.

Thanks for spotting this!

If you want your resource to return a value that can support mutation, you can return a JavaScript object with accessors and methods that can be used to mutate the value.

This is an advanced use-case because you will need to think about how external mutations should affect the running process. 

</details>

## A Ticking Stopwatch

> Here's a demo of a `Stapwatch` resource, similar to the above demo.

Reviewer: ```suggestion
Here's a demo of a `Stopwatch` resource, similar to the above demo.
```

Author: thanks!

The main difference here is that the return value is a function.

```js 
import { cell, resource } from "@starbeam/universal";

const formatter = new Intl.DateTimeFormat("en-US", {
  hour: "numeric",
  minute: "numeric",
  second: "numeric",
  hour12: false,
});

export const Stopwatch = resource((r) => {
  const time = cell(new Date());

  const interval = setInterval(() => {
    time.set(new Date());
  }, 1000);

  r.on.cleanup(() => {
    clearInterval(interval);
  });

  return () => {
    const now = time.current;

    return formatter.format(now);
  };
});
```

> To see this code in action, [checkout the live demo](https://limber.glimdown.com/edit?c=MQAggiDKAuD2AOB3AhtAxgCwFBbbAJgKYgBmhAdmsdMgOYDOIay5IARsQK72H4hwh8ASwBOhNNAA2ATxBjyREU1gBbeLHIVo9ADRZkjRIUmSQBuRUX8MxPERD1yQ%2BPELRrxSUIButgn5VCRhIRVQA6HAB5eGghDXoALhwAWhAAAy9fNJBUgH15RUYhVnhJZCpcrFSMn2J4MW8hQkRs-MtCESKSsqodEFLy4toPB1hOESozEmgOkcziAo7K6vn%2BhqbEdmNYFpA2hQ6u-p7CPoG0IbnaiwOlDkkdkfoxicJKrDTP2gArRnmsIRqWAidwAbwsz3GvSYxlMAF9SKEVCAAOSEFQcETJMSQ170FEAbhweHI9HcJGBKlQMyUAF4QJpNgBJchSMIAEVQhAAKoDCAAxSmoAAUACIKMkAKqQUV9UFYEAgDAvBIgUXkTiBERCNCyhUgFTFTgzVXqzUdHV6xU8En4U0arWWvSK5XjACMACZVSRkJIeHo4QBKIm4eLuGAIFDoDAgek4l5UYXC8EaEBB2MAPhA8sVJLJ-D5sZhJmFjJAnJmwsDwf1efcxRp3l9RZ40BZjd9ScDmYLgTCaHG8nc9LLFcIVb6boADDOa-qNP3JIQWJx4F2e2gl8gRO2Ok3JMKG3vfdWQ4qxNBxqwqz2c4rlKT3ORHvTYn2ByIh2f7xer6QhdANJhBSIhUtApY7DWipwlgQYhgAPDMahlDMGb6ky9aJCAiF8hmoKghGSCoJgcJwvBAD0b6EGhFFIQMqEfJ8QA&format=glimdown).

Reviewer: This one works, but I still don't understand the options.

You'd be right!

```js 
const formatter = new Intl.DateTimeFormat("en-US", {
  hour: "numeric",
  minute: "numeric",
  second: "numeric",
  hour12: false,
});
 
const Stopwatch = resource(({ use }) => {
  const time = use(Now);
         
  return () => formatter.format(time.current);
});
```

<details><summary>In Starbeam</summary>

```js 
const formatter = new Intl.DateTimeFormat("en-US", {
  hour: "numeric",
  minute: "numeric",
  second: "numeric",
  hour12: false,
});
 
const Stopwatch = Resource(({ use }) => {
  const time = use(Now);
         
  return Formula(() => formatter.format(time.current));
});
```

</details>

The `Stopwatch` resource instantiated a `Now` resource using its use method. That automatically links the `Now` instance to the owner of the `Stopwatch`, which means that when the component that instantiated the stopwatch is unmounted, the interval will be cleaned up.

## Using a Resource to Represent an Open Channel

Resources can do more than represent data like a ticking clock. You can use a resource with any long-running process, as long as you can represent it meaningfully as a "current value".

> <details><summany>Compared to other systems: Destiny of Unused Values</summary>

Reviewer: Formatting seems off here.

Author: in what way?

<details><summary>In Starbeam</summary>

```js 
import { Resource, Cell, Formula } from '@starbeam/universal';

function ChannelResource(channelName) {
  return Resource(({ on }) => {
    const lastMessage = Cell(null);
 
    const channel = Channel.subscribe(channelName.read());
 
    channel.onMessage((message) => {
      lastMessage.set(message);
    });
 
    on.cleanup(() => {
      channel.unsubscribe();
    });
 
    return Formula(() => {
      const prefix = `[${channelName.read()}] `;
      if (lastMessage.current === null) {
        return `${prefix} No messages received yet`;
      } /*E1*/ else {
        return `${prefix} ${lastMessage.current}`;
      }
    });
  });
}
```

</details>

`ChannelResource` is a JavaScript function that takes the channel name as a reactive input and returns a resource constructor.

> That resource constructor starts by subscribing to the current value of the `channelName`, and then telling ember to unsubscribe from the channel when the resource is cleaned up.

Reviewer: ```suggestion
That resource constructor starts by subscribing to the current value of the `channelName`, and then telling Ember to unsubscribe from the channel when the resource is cleaned up.
```

Author: Thanks!
